pkg/osv/osv.go

```diff
@@ -185,23 +187,74 @@ func Get(id string) (*models.Vulnerability, error) {
 	return &vuln, nil
 }
 
+type hydrateResponse struct {
+	batchIdx  int
+	resultIdx int
+	vuln      *models.Vulnerability
+	err       error
+}
+
 // Hydrate fills the results of the batched response with the full
 // Vulnerability details.
 func Hydrate(resp *BatchedResponse) (*HydratedBatchedResponse, error) {
-	// TODO(ochang): Parallelize requests, or implement batch GET.
 	hydrated := HydratedBatchedResponse{}
 
+	responseChan := make(chan hydrateResponse, maxConcurrentRequests)
+	rateLimiter := make(chan struct{}, maxConcurrentRequests)
+
+	// Run in separate goroutine so that we can immediately start
+	// listening to the responseChan, and it will be not blocked
+	go func(resp *BatchedResponse) {
+		for batchIdx, response := range resp.Results {
+			for resultIdx, vuln := range response.Vulns {
+				rateLimiter <- struct{}{}
+				go func(id string, batchIdx int, resultIdx int) {
+					vuln, err := Get(id)
+					if err != nil {
+						responseChan <- hydrateResponse{
```

Comment:
I don't think we need this channel. 

Can we just pre-allocate the hydrated structure (i.e. all indexes are accessible) and have each "worker" write to the specified index directly? 

I believe this is safe, per the example in https://pkg.go.dev/golang.org/x/sync/semaphore#example-package-WorkerPool:

```golang
	// Compute the output using up to maxWorkers goroutines at a time.
	for i := range out {
		// When maxWorkers goroutines are in flight, Acquire blocks until one of the
		// workers finishes.
		if err := sem.Acquire(ctx, 1); err != nil {
			log.Printf("Failed to acquire semaphore: %v", err)
			break
		}

		go func(i int) {
			defer sem.Release(1)
			out[i] = collatzSteps(i + 1)
		}(i)
	}
```

And this: https://go.dev/ref/spec#Variables:

> Structured variables of [array](https://go.dev/ref/spec#Array_types), [slice](https://go.dev/ref/spec#Slice_types), and [struct](https://go.dev/ref/spec#Struct_types) types have elements and fields that may be [addressed](https://go.dev/ref/spec#Address_operators) individually. Each such element acts like a variable.

And all the nested structures in `HydratedBatchedResponse` are slices.

Reply:
Good idea, done